File: scripts/analyze-amex-data.py
import csv
import requests
from datetime import datetime

def fetch_and_analyze_amex_data():
    """Fetch and analyze the American Express CSV data to identify processing issues"""
    
    # Fetch the American Express data
    amex_url = "https://hebbkx1anhila5yf.public.blob.vercel-storage.com/activity-REtDnwOPsQMKz1iyyRJhXrrbkwrnvE.csv"
    
    try:
        response = requests.get(amex_url)
        response.raise_for_status()
        
        # Parse CSV content
        csv_content = response.text
        lines = csv_content.strip().split('\n')
        
        print("=== AMERICAN EXPRESS DATA ANALYSIS ===")
        print(f"Total lines in file: {len(lines)}")
        
        # Parse CSV manually to handle potential issues
        rows = []
        for i, line in enumerate(lines):
            # Split by comma, handling quoted fields
            row = []
            current_field = ""
            in_quotes = False
            
            for char in line:
                if char == '"':
                    in_quotes = not in_quotes
                elif char == ',' and not in_quotes:
                    row.append(current_field.strip())
                    current_field = ""
                else:
                    current_field += char
            
            # Add the last field
            row.append(current_field.strip())
            rows.append(row)
        
        print(f"Parsed rows: {len(rows)}")
        
        # Analyze header
        if rows:
            header = rows[0]
            print(f"\nHeader row ({len(header)} columns):")
            for i, col in enumerate(header):
                print(f"  {i}: '{col}'")
        
        # Analyze first few data rows
        print(f"\nFirst 5 data rows analysis:")
        for i in range(1, min(6, len(rows))):
            row = rows[i]
            print(f"\nRow {i} ({len(row)} columns):")
            for j, cell in enumerate(row):
                if j < 10:  # Show first 10 columns
                    print(f"  [{j}] '{cell}'")
        
        # Analyze date column specifically
        print(f"\n=== DATE ANALYSIS ===")
        date_column = 0  # Column A
        date_formats_found = {}
        empty_dates = 0
        valid_dates = 0
        
        for i in range(1, len(rows)):  # Skip header
            if i >= len(rows):
                break
                
            row = rows[i]
            if len(row) > date_column:
                date_str = row[date_column].strip()
                
                if not date_str:
                    empty_dates += 1
                else:
                    # Try to identify date format
                    if '/' in date_str:
                        parts = date_str.split('/')
                        if len(parts) == 3:
                            format_key = f"{len(parts[0])}/{len(parts[1])}/{len(parts[2])}"
                            date_formats_found[format_key] = date_formats_found.get(format_key, 0) + 1
                            valid_dates += 1
                    elif '.' in date_str:
                        parts = date_str.split('.')
                        if len(parts) == 3:
                            format_key = f"{len(parts[0])}.{len(parts[1])}.{len(parts[2])}"
                            date_formats_found[format_key] = date_formats_found.get(format_key, 0) + 1
                            valid_dates += 1
        
        print(f"Empty dates: {empty_dates}")
        print(f"Valid dates: {valid_dates}")
        print(f"Date formats found: {date_formats_found}")
        
        # Analyze amount column
        print(f"\n=== AMOUNT ANALYSIS ===")
        amount_column = 4  # Column E (Betrag)
        empty_amounts = 0
        valid_amounts = 0
        negative_amounts = 0
        positive_amounts = 0
        amount_samples = []
        
        for i in range(1, min(50, len(rows))):  # Check first 50 rows
            row = rows[i]
            if len(row) > amount_column:
                amount_str = row[amount_column].strip()
                
                if not amount_str or amount_str == '0' or amount_str == '0,00':
                    empty_amounts += 1
                else:
                    valid_amounts += 1
                    amount_samples.append(amount_str)
                    
                    if '-' in amount_str or amount_str.startswith('('):
                        negative_amounts += 1
                    else:
                        positive_amounts += 1
        
        print(f"Empty/zero amounts: {empty_amounts}")
        print(f"Valid amounts: {valid_amounts}")
        print(f"Negative amounts: {negative_amounts}")
        print(f"Positive amounts: {positive_amounts}")
        print(f"Amount samples: {amount_samples[:10]}")
        
        # Analyze description columns
        print(f"\n=== DESCRIPTION ANALYSIS ===")
        desc_columns = [1, 6, 7]  # B, G, H
        
        for col_idx in desc_columns:
            non_empty = 0
            samples = []
            
            for i in range(1, min(20, len(rows))):
                row = rows[i]
                if len(row) > col_idx:
                    desc = row[col_idx].strip()
                    if desc:
                        non_empty += 1
                        if len(samples) < 5:
                            samples.append(desc)
            
            print(f"Column {col_idx}: {non_empty} non-empty, samples: {samples}")
        
        # Check for completely empty rows
        print(f"\n=== ROW COMPLETENESS ANALYSIS ===")
        completely_empty = 0
        mostly_empty = 0
        has_data = 0
        
        for i in range(1, len(rows)):
            row = rows[i]
            non_empty_cells = sum(1 for cell in row if cell.strip())
            
            if non_empty_cells == 0:
                completely_empty += 1
            elif non_empty_cells <= 2:
                mostly_empty += 1
            else:
                has_data += 1
        
        print(f"Completely empty rows: {completely_empty}")
        print(f"Mostly empty rows (≤2 cells): {mostly_empty}")
        print(f"Rows with substantial data: {has_data}")
        
        # Simulate current processing logic
        print(f"\n=== SIMULATING CURRENT PROCESSING LOGIC ===")
        processed = 0
        skipped_no_date = 0
        skipped_no_data = 0
        skipped_other = 0
        
        for i in range(1, len(rows)):  # Skip header
            row = rows[i]
            
            if len(row) < 2:
                skipped_other += 1
                continue
            
            # Check date
            date_str = row[0].strip() if len(row) > 0 else ""
            has_valid_date = bool(date_str and ('/' in date_str or '.' in date_str))
            
            # Check description
            desc = ""
            if len(row) > 1:
                desc = row[1].strip()
            if not desc and len(row) > 6:
                desc = row[6].strip()
            
            # Check amount
            amount_str = row[4].strip() if len(row) > 4 else ""
            has_amount = bool(amount_str and amount_str != '0' and amount_str != '0,00')
            
            # Current logic: needs date OR description OR amount
            if has_valid_date or desc or has_amount:
                processed += 1
            elif not has_valid_date:
                skipped_no_date += 1
            else:
                skipped_no_data += 1
        
        print(f"Would process: {processed}")
        print(f"Would skip (no date): {skipped_no_date}")
        print(f"Would skip (no data): {skipped_no_data}")
        print(f"Would skip (other): {skipped_other}")
        
        return {
            'total_rows': len(rows),
            'data_rows': len(rows) - 1,
            'date_formats': date_formats_found,
            'processing_estimate': processed
        }
        
    except Exception as e:
        print(f"Error analyzing Amex data: {e}")
        return None

# Run the analysis
result = fetch_and_analyze_amex_data()

File: scripts/analyze-amex-data.ts
// Analyze the American Express input and output files to identify processing issues

async function analyzeAmexData() {
  console.log("=== ANALYZING AMERICAN EXPRESS DATA ===\n")

  try {
    // Fetch the input file
    console.log("Fetching input file...")
    const inputResponse = await fetch(
      "https://hebbkx1anhila5yf.public.blob.vercel-storage.com/activity-QY4bQye0M8zeJq1r97LKbVTIf0ZxKb.csv",
    )
    const inputText = await inputResponse.text()

    // Fetch the output file
    console.log("Fetching output file...")
    const outputResponse = await fetch(
      "https://hebbkx1anhila5yf.public.blob.vercel-storage.com/converted_amex_2025-06-30-3TKhXyR7Wk6ywdJ8qs9BVyD7DWTqnF.csv",
    )
    const outputText = await outputResponse.text()

    // Parse input CSV
    const inputLines = inputText.split("\n").filter((line) => line.trim() !== "")
    const outputLines = outputText.split("\n").filter((line) => line.trim() !== "")

    console.log(`Input file: ${inputLines.length} lines`)
    console.log(`Output file: ${outputLines.length} lines`)

    // Analyze input structure
    console.log("\n=== INPUT FILE ANALYSIS ===")
    console.log("First 5 lines of input:")
    inputLines.slice(0, 5).forEach((line, i) => {
      const columns = line.split(",")
      console.log(`Line ${i + 1}: ${columns.length} columns`)
      console.log(`  Content: ${line.substring(0, 100)}...`)
    })

    // Check for header
    const headerLine = inputLines[0]
    const headerColumns = headerLine.split(",")
    console.log(`\nHeader columns (${headerColumns.length}):`)
    headerColumns.forEach((col, i) => {
      console.log(`  ${i}: "${col.trim()}"`)
    })

    // Analyze data rows
    console.log("\n=== DATA ROWS ANALYSIS ===")
    const dataRows = inputLines.slice(1) // Skip header
    console.log(`Total data rows: ${dataRows.length}`)

    // Sample data analysis
    console.log("\nFirst 10 data rows analysis:")
    dataRows.slice(0, 10).forEach((line, i) => {
      const columns = line.split(",")
      console.log(`Row ${i + 2}: ${columns.length} columns`)
      console.log(`  Date (col 0): "${columns[0]?.trim()}"`)
      console.log(`  Description (col 1): "${columns[1]?.trim()}"`)
      console.log(`  Amount (col 4): "${columns[4]?.trim()}"`)
      console.log(`  Statement desc (col 6): "${columns[6]?.trim()}"`)
      console.log("---")
    })

    // Check for empty or problematic rows
    let emptyRows = 0
    let shortRows = 0
    let validRows = 0

    dataRows.forEach((line, i) => {
      const columns = line.split(",")
      if (line.trim() === "") {
        emptyRows++
      } else if (columns.length < 5) {
        shortRows++
        console.log(`Short row ${i + 2}: ${columns.length} columns - "${line}"`)
      } else {
        validRows++
      }
    })

    console.log(`\n=== ROW QUALITY ANALYSIS ===`)
    console.log(`Empty rows: ${emptyRows}`)
    console.log(`Short rows (< 5 columns): ${shortRows}`)
    console.log(`Valid rows: ${validRows}`)
    console.log(`Expected processed: ${validRows}`)

    // Analyze date formats
    console.log("\n=== DATE FORMAT ANALYSIS ===")
    const dateFormats = new Map()
    dataRows.slice(0, 50).forEach((line, i) => {
      const columns = line.split(",")
      const dateStr = columns[0]?.trim()
      if (dateStr) {
        const format = analyzeDateFormat(dateStr)
        dateFormats.set(format, (dateFormats.get(format) || 0) + 1)
        if (i < 10) {
          console.log(`Row ${i + 2}: "${dateStr}" -> ${format}`)
        }
      }
    })

    console.log("\nDate format distribution:")
    dateFormats.forEach((count, format) => {
      console.log(`  ${format}: ${count} occurrences`)
    })

    // Analyze output file
    console.log("\n=== OUTPUT FILE ANALYSIS ===")
    console.log("First 5 lines of output:")
    outputLines.slice(0, 5).forEach((line, i) => {
      console.log(`Line ${i + 1}: ${line}`)
    })

    const outputDataRows = outputLines.slice(1) // Skip header
    console.log(`\nOutput data rows: ${outputDataRows.length}`)
    console.log(`Missing rows: ${dataRows.length - outputDataRows.length}`)
  } catch (error) {
    console.error("Error analyzing data:", error)
  }
}

function analyzeDateFormat(dateStr: string): string {
  if (!dateStr) return "empty"

  if (dateStr.match(/^\d{1,2}\/\d{1,2}\/\d{4}$/)) {
    return "DD/MM/YYYY or MM/DD/YYYY"
  } else if (dateStr.match(/^\d{1,2}\.\d{1,2}\.\d{4}$/)) {
    return "DD.MM.YYYY"
  } else if (dateStr.match(/^\d{4}-\d{2}-\d{2}$/)) {
    return "YYYY-MM-DD"
  } else {
    return `unknown: ${dateStr}`
  }
}

// Run the analysis
analyzeAmexData()
